internal/engine/crd_test.go

```diff
@@ -0,0 +1,271 @@
+// rather than the struct value itself.
+//
+// However, this _#def is ephemerally added only during printing. It does not
+// exist at runtime in the graph; do not look for it with e.g. LookupPath().
+var goldenBucketFirstSchema = `import "strings"
+
+_#def
+_#def: {
+	// APIVersion defines the versioned schema of this representation
+	// of an object. Servers should convert recognized schemas to the
+	// latest internal value, and may reject unrecognized values.
+	// More info:
+	// https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+	apiVersion: "source.toolkit.fluxcd.io/v1beta1"
+
+	// Kind is a string value representing the REST resource this
+	// object represents. Servers may infer this from the endpoint
+	// the client submits requests to. Cannot be updated. In
+	// CamelCase. More info:
+	// https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	kind: "Bucket"
+	metadata: {
+		name:      string
+		namespace: string
+		labels?: {
+			[string]: string
+		}
+		annotations?: {
+			[string]: string
+		}
+	}
+
+	// BucketSpec defines the desired state of an S3 compatible bucket
+	spec?: {
+		// AccessFrom defines an Access Control List for allowing
+		// cross-namespace references to this object.
+		accessFrom?: {
+			// NamespaceSelectors is the list of namespace selectors to which
+			// this ACL applies. Items in this list are evaluated using a
+			// logical OR operation.
+			namespaceSelectors: [...{
+				// MatchLabels is a map of {key,value} pairs. A single {key,value}
+				// in the matchLabels map is equivalent to an element of
+				// matchExpressions, whose key field is "key", the operator is
+				// "In", and the values array contains only "value". The
+				// requirements are ANDed.
+				matchLabels?: {
+					[string]: string
+				}
+				...
+			}]
+			...
+		}
+
+		// The bucket name.
+		bucketName: string
+
+		// The bucket endpoint address.
+		endpoint: string
+
+		// Ignore overrides the set of excluded patterns in the
+		// .sourceignore format (which is the same as .gitignore). If not
+		// provided, a default will be used, consult the documentation
+		// for your version to find out what those are.
+		ignore?: string
+
+		// Insecure allows connecting to a non-TLS S3 HTTP endpoint.
+		insecure?: bool
+
+		// The interval at which to check for bucket updates.
+		interval: string
+
+		// The S3 compatible storage provider name, default ('generic').
+		provider?: "generic" | "aws" | "gcp" | *"generic"
+
+		// The bucket region.
+		region?: string
+
+		// The name of the secret containing authentication credentials
+		// for the Bucket.
+		secretRef?: {
+			// Name of the referent.
+			name: string
+			...
+		}
+
+		// This flag tells the controller to suspend the reconciliation of
+		// this source.
+		suspend?: bool
+
+		// The timeout for download operations, defaults to 60s.
+		timeout?: string | *"60s"
+		...
```

Comment:
Can we make all definitions as closed? You shouldn’t be able to add fields to spec, undefined fields are rejected by Kubernetes API.

Reply:
Yeah, i fixed this in my latest push. There's a TODO there around properly handling the exception cases where unspecified fields are allowed. But i think it's good enough for an initial implementation, at least. Openness/closedness is a surprisingly deep rabbit hole.

Reply:
Oh yeah I totally forgot about opened fields in CRDs, they do have a special annotation: https://github.com/fluxcd/helm-controller/blob/a8c15d3700afce4309c12f37803f96ed166991b5/config/crd/bases/helm.toolkit.fluxcd.io_helmreleases.yaml#L747

Reply:
Yeah, that's the one, `x-kubernetes-preserve-unknown-fields`. From [the docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.28/#jsonschemaprops-v1-apiextensions-k8s-io):

> `x-kubernetes-preserve-unknown-fields` stops the API server decoding step from pruning fields which are not specified in the validation schema. This affects fields recursively, but switches back to normal pruning behaviour if nested properties or additionalProperties are specified in the schema. This can either be true or undefined. False is forbidden.

IMO, this is most properly handled in `cuelang.org/go/encoding/openapi.Extract` itself. By the time we have a `cue.Value` out, the custom k8s props have been lost. It may be possible to do a hack here where we walk the AST of the original openapi in order to figure exactly which levels should be open vs. closed, but i can't currently imagine a way of doing that without a bunch of painful wheel reinvention.

I don't have a good sense of how harmful it would be to not support this property from the beginning. Would it be acceptable to ignore it for now and treat this PR as an initial pass, or would it impact most users in a way that is likely to make the feature generally unusable?

Reply:
I would say it's unusable without handling `x-kubernetes-preserve-unknown-fields`, lots of CRDs have it Flux, cert-manager, etc. As a workaround, we could add an arg e,g, `closed` so users could set it to false and then all definitions are left opened, you lose the validation but at least you can generate valid objects.